### ISSUE TITLE
Remote plugin with conditional loading is not handled well

### ISSUE THREAD

**Issue by jdhao** [wbthomason/packer.nvim#464]
- `nvim --version`: Latest
- `git --version`: 2.21.0
- Operating system/version: macOS
- Terminal name/version: kitty

### Steps to reproduce

```
lua << EOF
require('packer').startup(function()
    use {'numirias/semshi', run = ':UpdateRemotePlugins', ft = 'python'}
end)
EOF
```

1. Use the above config to start nvim: `nvim -u init.vim`.
2. Use PakcerCompile to generate packer_compiled.lua.
3. Run `PackerUpdate` to update the plugin.
4. Open a Python source file.

### Actual behaviour

We see error that command `Semshi enable` is not available.

### Expected behaviour

No error occurs.

### Analysis of the error

If Semshi is set to opt plugin, when we run `:PackerUpdate`, the command `:
UpdateRemotePlugins` will be run. At this time, since Semshi is only loaded
when we are editing a Python source file, it is not in the runtimepath. So its
manifest will not be generated in rplugin.vim (located under
~/.local/share/nvim/rplugin.vim). Thus, when we open a Python file afterwards,
although Semshi is in runtimepath, but its manifest is missing, we see error
that command `Semshi enable` is not found.

After loading this plugin, i.e., when we open a Python source file,
we need to run `:UpdateRemotePlugins` manually so that semshi can work correctly.

Maybe the user should be warned if they try to conditionaly load a nvim remote plugin.

**Comment by wbthomason** [comment]
Ah, yes. I've run into this before. We're supposed to handle this case already by loading the plugin before executing its `run` key. However, I think we don't do anything with `rplugin` directories, which would cause this behavior.

Just to check: does the `Semshi` command get created as expected if you do not lazy-load the plugin? Assuming so, I think there's an easy fix here which will make things work as expected with lazy-loading.

**Comment by jdhao** [comment]
Yes, without conditional loading, Semshi works when we open a Python file:

```
lua << EOF
require('packer').startup(function()
  use {'numirias/semshi', run = ':UpdateRemotePlugins'}
end)
EOF
```

**Comment by wbthomason** [comment]
Thanks. I'm actually having trouble reproducing this bug on my end - the "easy fix" I thought we could make is already in place (`packadd`ing `opt` plugins before executing their `run` keys), and - with the exact spec you gave in your initial report - I do get a `Semshi` command correctly generated.

**Comment by jdhao** [comment]
What if you delete `~/.local/share/nvim/rplugin.vim` and run the step in the initial report. In my case, there is no command related to Semshi in the generated rplugin.vim file.

**Comment by wbthomason** [comment]
I still get a `Semshi` command. I think what's happening here is actually that I didn't have `semshi` installed before trying to reproduce this - so the post-install/update hook (i.e. `:UpdateRemotePlugins`) runs as expected for me, with `semshi` loaded. I'm guessing you installed it, then added the `run` key? `run` keys are only executed on plugin update or installation, which would explain why you don't get the command.

I'd guess that you *do* get the command when you switch to not lazy-loading because that constitutes an "update" (`packer` moves the plugin's directory), and so the hook executes.

**Comment by akinsho** [comment]
I think I might be hitting something similar with `wilder.nvim` I seemed to be getting errors loading the plugin on `CmdlineEnter`, it's a remote plugin and on occasion it errors about it's missing `init` function which can I'm able to resolve by manually running `:UpdateRemotePlugins`. Looking at the code it seems this should be being run on install. Not sure if it runs on update, I'm not super familiar with how packer behaves in this scenario, or if this is exactly the same issue tbh.

1.) is a run key required or expected to be manually added by a user for remote plugins?
2.) is there a reason why updating a plugin might not trigger the `UpdateRemotePlugins` command (note: I did have an error during an update before this error re-appeared, but the first time it appeared in a similar way to this)

**Comment by wbthomason** [comment]
We should run `UpdateRemotePlugins` automatically and unconditionally in all of `install`, `update`, and `sync`. I think the problem is that (line 214 of `plugin_utils`) we don't load `opt` plugins. This is normally what we want to prevent plugins which should only load under certain conditions from being arbitrarily loaded after an update, but it misses this case.

I think we might need to add a check to manually look for and source any `rplugin` files before running `UpdateRemotePlugins` to fix this. I'm not sure why I couldn't reproduce it before.

**Comment by akinsho** [comment]
@wbthomason I think I'm a little confused, do you mean search for the `rplugin.vim` manifest file and manually source it, except I presume it won't be there or will be wrong since that's what the `UpdateRemotePlugins` command does, I've looked through the wilder plugin as an example, and it's `rplugin` directory is just full of python code, so we couldn't load that. Not sure what would need sourcing?

I'm guessing you meant, we need to actually load remote plugins, since otherwise the rplugin manifest won't have picked up their code and would be wrong? Or could we run the update rplugins function just before loading a conditional remote plugin? 🤔

**Comment by wbthomason** [comment]
So, I think the set of paths that `UpdateRemotePlugins` searches does *not* include `opt` plugins which aren't loaded. I was referring to manually doing that part of the remote plugin update for opt plugins, or otherwise ensuring that they get picked up for `UpdateRemotePlugins`. I suppose running `UpdateRemotePlugins` just after `packadd` in the loader logic could also work?

**Comment by akinsho** [comment]
Ah, I'd slightly gone down the rabbit hole of checking if a plugin is an rplugin and then loading it at the check you mentioned in the `plugin_utils` file, that way they are loaded and update remote plugins works. I could change that to just call update remote plugins after every packadd 🤷🏿 I'll open what I've got as a draft PR and you can have a look. Will tweak it as time permits

**Comment by amenonsen** [comment]
I'm running into this problem with [vim-ultest](https://github.com/rcarriga/vim-ultest). I tried to do this in the `run` function as a workaround:

```lua
    -- Integrates with vim-test and nvim-dap to run tests.
    use {
        { 'vim-test/vim-test', ft = { 'python' } },
        {
            "rcarriga/vim-ultest", ft = { 'python' },
            after = { "vim-test" },
            run = function ()
                -- vim.cmd[[packadd vim-test vim-ultest]]
                vim.cmd[[PackerLoad vim-test vim-ultest]]
                vim.cmd[[UpdateRemotePlugins]]
            end, …
        }
    }
```

It works if I run PackerSync and then run `PackerLoad vim-test vim-ultest` by hand and then run `UpdateRemotePlugins`, but my run function above does not work. I tried to remove the `after` specification to see if that helped, but it didn't, and neither did using packadd instead of PackerLoad. (Of course, it works fine without setting `ft`.) I'm not sure how to debug this, but I'm happy to try if someone can point me in the right direction.
